numpy 1.12.0 seems to break f2py

I now get lots of output of the form

```
(...)
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c: At top level:
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:1052:21: error: expected expression before â,â token
   {"namezone",2,{{++,8}},NPY_CHAR},
                     ^
//tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2468:15: note: (near initialization for âf2py_vars_def[848].dims.d[0]â)
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2474:18: error: expected expression before â,â token
   {"ionbmax",2,{{,8}},NPY_CHAR},
                  ^
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2474:18: error: initializer element is not constant
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2474:18: note: (near initialization for âf2py_vars_def[854].dims.d[0]â)
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2501:15: error: invalid type argument of unary â*â (have âintâ)
   {"dbuf",1,{{*(+2)}},NPY_DOUBLE},
               ^~~~~
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2501:15: error: initializer element is not constant
   {"dbuf",1,{{*(+2)}},NPY_DOUBLE},
               ^
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2501:15: note: (near initialization for âf2py_vars_def[881].dims.d[0]â)
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:2503:18: error: expected expression before â}â token
   {"ppnold",1,{{*}},NPY_DOUBLE},
                  ^
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:1667:19: error: wrong type argument to unary minus
   {"idumhead",1,{{-}},NPY_INT},
                   ^
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:3547:14: note: (near initialization for âf2py_zbrn_def[13].dims.d[0]â)
/tmp/tmp8i0j4nc_/src.linux-x86_64-3.6/_keplermodule.c:3699:19: error: expected expression before â,â token
   {"uuidhist",3,{{,,16}},NPY_CHAR},
                   ^
(...)
error: Command "gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/tmp/tmpgnfiujz_/src.linux-x86_64-3.6 -I/home/alex/Python/lib/python3.6/site-packages/numpy/core/include -I/home/alex/Python/include/python3.6m -c /tmp/tmpgnfiujz_/src.linux-x86_64-3.6/_keplermodule.c -o /tmp/tmpgnfiujz_/tmp/tmpgnfiujz_/src.linux-x86_64-3.6/_keplermodule.o" failed with exit status 1
```

in f2py in NumPy version 1.12.0 from PyPi and I can no longer compile the code.  For comparison, all worked fine in NumPy 1.11.3.

I am using Python 3.6.0, compiled with optimisation from the sources on Python.org on Fedora 25, gcc version 6.3.1 20161221 (Red Hat 6.3.1-1).

Seems to me something in the parsing is no longer working, not sure about the origin of all the special characters in the output either.

--- a/numpy/f2py/crackfortran.py
+++ b/numpy/f2py/crackfortran.py
@@ -2451,7 +2451,7 @@ def get_parameters(vars, global_params={}):
                 if not selected_kind_re.match(v):
                     v_ = v.split('_')
                     # In case there are additive parameters
-                    if len(v_)>1: 
+                    if len(v_) > 1: 
                         v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
 
             # Currently this will not work for complex numbers.

--- a/numpy/f2py/crackfortran.py
+++ b/numpy/f2py/crackfortran.py
@@ -2451,7 +2451,8 @@ def get_parameters(vars, global_params={}):
                 if not selected_kind_re.match(v):
                     v_ = v.split('_')
                     # In case there are additive parameters
-                    v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
+                    if len(v_)>1: 
+                        v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
 
             # Currently this will not work for complex numbers.
             # There is missing code for extracting a complex number,

--- a/numpy/f2py/tests/src/parameter/constant_compound.f90
+++ b/numpy/f2py/tests/src/parameter/constant_compound.f90
@@ -1,0 +1,15 @@
+! Check that parameters are correct intercepted.
+! Constants with comma separations are commonly
+! used, for instance Pi = 3._dp
+subroutine foo_compound_int(x)
+  implicit none
+  integer, parameter :: ii = selected_int_kind(9)
+  integer(ii), intent(inout) :: x
+  dimension x(3)
+  integer(ii), parameter :: three = 3_ii
+  integer(ii), parameter :: two = 2_ii
+  integer(ii), parameter :: six = three * 1_ii * two
+
+  x(1) = x(1) + x(2) + x(3) * six
+  return
+end subroutine

--- a/numpy/f2py/tests/src/parameter/constant_non_compound.f90
+++ b/numpy/f2py/tests/src/parameter/constant_non_compound.f90
@@ -1,0 +1,23 @@
+! Check that parameters are correct intercepted.
+! Specifically that types of constants without 
+! compound kind specs are correctly inferred
+! adapted Gibbs iteration code from pymc 
+! for this test case 
+subroutine foo_non_compound_int(x)
+  implicit none
+  integer, parameter :: ii = selected_int_kind(9)
+
+  integer(ii)   maxiterates
+  parameter (maxiterates=2)
+
+  integer(ii)   maxseries
+  parameter (maxseries=2)
+
+  integer(ii)   wasize
+  parameter (wasize=maxiterates*maxseries)
+  integer(ii), intent(inout) :: x
+  dimension x(wasize)
+
+  x(1) = x(1) + x(2) + x(3) + x(4) * wasize
+  return
+end subroutine

--- a/numpy/f2py/tests/test_parameter.py
+++ b/numpy/f2py/tests/test_parameter.py
@@ -19,6 +19,7 @@ class TestParameters(util.F2PyTest):
                _path('src', 'parameter', 'constant_integer.f90'),
                _path('src', 'parameter', 'constant_both.f90'),
                _path('src', 'parameter', 'constant_compound.f90'),
+               _path('src', 'parameter', 'constant_non_compound.f90'),
     ]
 
     @dec.slow
@@ -53,6 +54,13 @@ class TestParameters(util.F2PyTest):
         x = np.arange(3, dtype=np.int32)
         self.module.foo_compound_int(x)
         assert_equal(x, [0 + 1 + 2*6, 1, 2])
+
+    @dec.slow
+    def test_constant_non_compound_int(self):
+        # check values
+        x = np.arange(4, dtype=np.int32)
+        self.module.foo_non_compound_int(x)
+        assert_equal(x, [0 + 1 + 2 + 3*4, 1, 2, 3])
 
     @dec.slow
     def test_constant_integer_int(self):

--- a/numpy/f2py/tests/test_parameter.py
+++ b/numpy/f2py/tests/test_parameter.py
@@ -18,6 +18,7 @@ class TestParameters(util.F2PyTest):
     sources = [_path('src', 'parameter', 'constant_real.f90'),
                _path('src', 'parameter', 'constant_integer.f90'),
                _path('src', 'parameter', 'constant_both.f90'),
+               _path('src', 'parameter', 'constant_compound.f90'),
     ]
 
     @dec.slow
@@ -41,6 +42,17 @@ class TestParameters(util.F2PyTest):
         x = np.arange(3, dtype=np.float64)
         self.module.foo_double(x)
         assert_equal(x, [0 + 1 + 2*3, 1, 2])
+
+    @dec.slow
+    def test_constant_compound_int(self):
+        # non-contiguous should raise error
+        x = np.arange(6, dtype=np.int32)[::2]
+        assert_raises(ValueError, self.module.foo_compound_int, x)
+
+        # check values with contiguous array
+        x = np.arange(3, dtype=np.int32)
+        self.module.foo_compound_int(x)
+        assert_equal(x, [0 + 1 + 2*6, 1, 2])
 
     @dec.slow
     def test_constant_integer_int(self):

